### PR TITLE
Book CRUD

### PR DESCRIPTION
**Pull request by anaturrillo**
# Summary
Creates CRUD operations for book controller and custom book exceptions.
I run into an issue with Lombok. Using the getter and setter notations the received entity had all properties null. I added getters and setters and worked fine. I will retry to implement Lombok later.

# Trello card 
https://trello.com/c/KUWcHeSB/18-crear-las-operaciones-crud-de-book

### CODE REVIEW

File: src/main/java/wolox/training/controllers/BookController.java [hunk 1]
```diff
@@ -1,12 +1,65 @@
...
+@RequestMapping("/api/books")
 public class BookController {
+
+    @Autowired
+    private BookRepository bookRepository;
+
+    @GetMapping("/{author}")
+    public Book findByAuthor(@PathVariable String author) {
+        Book found = bookRepository.findFirstByAuthor(author);
+        if (found == null) throw new BookNotFoundException();
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/controllers/BookController.java:32]: If the repository method returned an optional, you could do `bookRepository.findFirstByAuthor(author).orElseThrow(...)`  specifying that the book was not found, thus avoiding the null check 😄👌

File: src/main/java/wolox/training/controllers/BookController.java [hunk 2]
```diff
@@ -1,12 +1,65 @@
...
+    public Book create(@RequestBody Book book) {
+        return bookRepository.save(book);
+    }
+
+    @PutMapping("/{id}")
+    public Book update(@RequestBody Book book, @PathVariable Long id) {
+        if (book.getId() != id) {
+            throw new BookIdMismatchException("The id does not match");
+        }
+        if (bookRepository.existsById(id)) {
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/controllers/BookController.java:47]: I think you should check the opposite (`!bookRepository.existsById(id)`)

File: src/main/java/wolox/training/controllers/BookController.java [hunk 3]
```diff
@@ -1,12 +1,65 @@
...
+            throw new BookNotFoundException();
+        }
+
+        return bookRepository.save(book);
+    }
+
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
+        if (bookRepository.existsById(id)) {
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/controllers/BookController.java:56]: Same here 😅

File: src/main/java/wolox/training/exceptions/BookIdMismatchException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package wolox.training.exceptions;
+
+public class BookIdMismatchException extends RuntimeException {
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/exceptions/BookIdMismatchException.java:3]: You can add a status code to the exceptions, check this out: https://www.baeldung.com/spring-response-status-exception 😄

**Review by Jimenarosello — CHANGES_REQUESTED** [review]
Just a few comments 😃

File: src/main/java/wolox/training/controllers/BookController.java (hunk 2 above)
Comment (anaturrillo) [reply src/main/java/wolox/training/controllers/BookController.java:47]: I realized it later :woman_facepalming:

File: src/main/java/wolox/training/controllers/BookController.java (hunk 3 above)
Comment (anaturrillo) [reply src/main/java/wolox/training/controllers/BookController.java:56]: :woman_facepalming:

**Pull request merged**